src/Microsoft.Health.Fhir.SqlServer/Features/Watchdogs/DefragWatchdog.cs

```diff
@@ -7,193 +7,302 @@
 
-        public void Start()
+        private readonly Func<IScoped<SqlConnectionWrapperFactory>> _sqlConnectionWrapperFactory;
+        private readonly SchemaInformation _schemaInformation;
+        private readonly Func<IScoped<SqlQueueClient>> _sqlQueueClient;
+        private readonly ILogger<DefragWatchdog> _logger;
+
+        private TimeSpan _timerDelay;
+        private bool _storageReady;
+
+        public DefragWatchdog(
+            Func<IScoped<SqlConnectionWrapperFactory>> sqlConnectionWrapperFactory,
+            SchemaInformation schemaInformation,
+            Func<IScoped<SqlQueueClient>> sqlQueueClient,
+            ILogger<DefragWatchdog> logger)
         {
-            InitParams();
-            StartTimer(_periodSec);
+            _schemaInformation = EnsureArg.IsNotNull(schemaInformation, nameof(schemaInformation));
+            _sqlConnectionWrapperFactory = EnsureArg.IsNotNull(sqlConnectionWrapperFactory, nameof(sqlConnectionWrapperFactory));
+            _sqlQueueClient = EnsureArg.IsNotNull(sqlQueueClient, nameof(sqlQueueClient));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
         }
 
-        public void Change(double periodSec)
+        internal void ChangeDelay(TimeSpan newTimerDelay)
         {
-            ChangeTimer(periodSec);
+            _timerDelay = newTimerDelay;
         }
 
-        protected override void Run()
+        public async Task Initialize(CancellationToken cancellationToken)
         {
-            if (_schemaInformation.Current < SchemaVersionConstants.Defrag || !IsEnabled())
+            if (!_storageReady || _schemaInformation.Current < SchemaVersionConstants.Defrag)
             {
-                return;
+                cancellationToken.ThrowIfCancellationRequested();
+                await Task.Delay(TimeSpan.FromMinutes(1), cancellationToken);
             }
 
-            try
+            await InitParamsAsync();
+
+            _threads = await GetThreads(cancellationToken);
+            _heartbeatPeriodSec = await GetHeartbeatPeriod(cancellationToken);
+            _heartbeatTimeoutSec = await GetHeartbeatTimeout(cancellationToken);
+            _periodHour = await GetPeriod(cancellationToken);
+
+            _timerDelay = TimeSpan.FromHours(_periodHour);
+        }
+
+        public async Task ExecuteAsync(CancellationToken cancellationToken)
+        {
+            bool initialRun = true;
+
+            while (!cancellationToken.IsCancellationRequested)
             {
-                var id = GetCoordinatorJob();
-                if (id.jobId == -1)
+                if (!initialRun)
                 {
-                    return;
+                    await Task.Delay(_timerDelay, cancellationToken);
                 }
 
-                ExecWithHeartbeat(
-                () =>
-                {
-                    try
-                    {
-                        InitDefrag(id.groupId);
+                initialRun = false;
```

Comment:
In my previous implementation I used small random delay to ensure that all watch dog processes from multiple VMs will not go to the database at once. And period was used as true period. Now we have only one delay. I think we should add small random delay back.